Attempted to add Anticipated Shows and Anticipated Movies sensors

mkanet opened the pull request:
I really appreciate your help to add `Anticipated Shows` and `Anticipated Movies` sensors.  Once I'm more familiar with the structure of your code, I will be able to submit PRs more easily.

**PS:** I will remove the comments I added soon.  I had it there temporarily so I can easily spot the code I added.

downey-lv commented on custom_components/trakt_tv/apis/trakt.py:
```diff
@@ -390,6 +390,59 @@ async def fetch_recommendations(self, configured_kinds: list[TraktKind]):
+
+    async def fetch_anticipated_shows(self, limit: int):
+        return await self.fetch_anticipated("shows", limit)
+
```
You can keep only fetch_anticipated and instead pass the path from KIND:
```
    async def fetch_anticipated(self, path: str, limit: int):
        return await self.request(
            "get", f"{path}?limit={limit}&ignore_collected=false"
        )
````

downey-lv replied:
New fetch_anticipated looks good, but remove fetch_anticipated_movies and fetch_anticipated_shows, you're not using them anymore

downey-lv commented on custom_components/trakt_tv/apis/trakt.py:
```diff
@@ -390,6 +390,59 @@ async def fetch_recommendations(self, configured_kinds: list[TraktKind]):
+                for kind in kinds
+            ]
+        )
+
```
After updating fetch_anticipated you can simplify this here:
```
data = await gather(
    *[
        self.fetch_anticipated(
            kind.value.path,
            configuration.get_anticipated_max_medias(kind.value.identifier),
        )
        for kind in kinds
    ]
)
```

downey-lv commented on custom_components/trakt_tv/apis/trakt.py:
```diff
@@ -390,6 +390,59 @@ async def fetch_recommendations(self, configured_kinds: list[TraktKind]):
+    # start of new code
+    async def fetch_anticipated(self, path: str, limit: int):
+        return await self.request(
+            "get", f"{path}/anticipated?limit={limit}&ignore_collected=false"
```
ignore_collected should probably be user's choice, i.e. set true or false from config

downey-lv commented on custom_components/trakt_tv/schema.py:
```diff
@@ -76,4 +79,16 @@ def recommendation_schema() -> Dict[str, Any]:
+# start of new code
+def anticipated_schema() -> Dict[str, Any]:
+    subschemas = {}
+    for trakt_kind in [TraktKind.ANTICIPATED_MOVIE, TraktKind.ANTICIPATED_SHOW]:
```
Re-use ANTICIPATED_KINDS here

downey-lv commented on custom_components/trakt_tv/sensor.py:
```diff
@@ -51,6 +51,21 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
             sensors.append(sensor)
 
+        # start of new code
+        if trakt_kind in [TraktKind.ANTICIPATED_MOVIE, TraktKind.ANTICIPATED_SHOW]:
```
Re-use ANTICIPATED_KINDS here

mkanet commented:
Hi, I tried my best to make the changes you requested.  See my second commit.  Please forgive me if I didn't do something correctly.  The good news is that I am not getting any error messages in home assistant log when adding this integration via config_flow.  

The bad news is that the new sensors are still not being created on my Home Assistant instance after adding this modified trakt integration via config_flow.  Only the existing sensors that were there before are being created.  

Below, is the configuration I am specifying in configuration.yaml.  
```
trakt_tv:
  language: en
  timezone: America/Los_Angeles
  sensors:
    upcoming:
      new_show:
        days_to_fetch: 60
        max_medias: 10
      movie:
        days_to_fetch: 60
        max_medias: 10
      dvd:
        days_to_fetch: 60
        max_medias: 10
    recommendation:
      show:
        max_medias: 10
      movie:
        max_medias: 10
    anticipated:
      anticipated_movie:
        max_medias: 10
      anticipated_show:
        max_medias: 10
```

Can you please tell me what I need to change or add to the existing source files so the integration will actually create the 2 new sensors:  `anticipated_movies` and `anticipated_shows`?  Thank you for all your help and patience!

downey-lv commented on custom_components/trakt_tv/apis/trakt.py:
```diff
@@ -327,7 +327,7 @@ async def fetch_next_to_watch(
 
     async def fetch_upcomings(
-        self, configured_kinds: list[TraktKind], all_medias: bool
+        self, configured_kinds: "list[TraktKind]", all_medias: bool
```
Revert this

downey-lv commented on custom_components/trakt_tv/apis/trakt.py:
```diff
@@ -353,7 +353,7 @@ async def fetch_recommendation(self, path: str, max_items: int):
         )
 
-    async def fetch_recommendations(self, configured_kinds: list[TraktKind]):
+    async def fetch_recommendations(self, configured_kinds: "list[TraktKind]"):
```
Revert this

downey-lv commented on custom_components/trakt_tv/apis/trakt.py:
```diff
@@ -390,6 +390,61 @@ async def fetch_recommendations(self, configured_kinds: list[TraktKind]):
+            TraktKind.ANTICIPATED_MOVIE,
+            TraktKind.ANTICIPATED_SHOW,
+        ]
+
```
Should be defined in kind.py and imported in this file

downey-lv replied:
Still an issue

downey-lv commented on custom_components/trakt_tv/schema.py:
```diff
@@ -8,6 +8,7 @@
 
 from .const import DOMAIN, LANGUAGE_CODES
 from .models.kind import BASIC_KINDS, NEXT_TO_WATCH_KINDS, TraktKind
+from .models.kind import ANTICIPATED_KINDS
```
Move ANTICIPATED_KINDS up to the same import as others from .models.kind. 
Also, you're importing them but haven't defined them there yet (see my previous comment on ANTICIPATED_KINDS)

downey-lv commented on custom_components/trakt_tv/sensor.py:
```diff
@@ -8,6 +8,7 @@
 from .configuration import Configuration
 from .const import DOMAIN
 from .models.kind import BASIC_KINDS, NEXT_TO_WATCH_KINDS, TraktKind
+from .models.kind import ANTICIPATED_KINDS
```
Same regarding import here

dylandoamaral commented:
> Thank you. I tried my best to make the changes you requested. **See my second commit: [c1dac64](https://github.com/dylandoamaral/trakt-integration/commit/c1dac6404c824d2db8b1599ee209d8f263874dd8).** Please forgive me if I didn't do something correctly. The good news is that I am not getting any error messages in home assistant log when adding this integration via config_flow.
> 
> The bad news is that the new sensors are still not being created on my Home Assistant instance after adding this modified trakt integration via config_flow. Only the existing sensors that were there before are being created.
> 
> Below, is the configuration I am specifying in configuration.yaml.
> 
> ```
> trakt_tv:
>   language: en
>   timezone: America/Los_Angeles
>   sensors:
>     upcoming:
>       new_show:
>         days_to_fetch: 60
>         max_medias: 10
>       movie:
>         days_to_fetch: 60
>         max_medias: 10
>       dvd:
>         days_to_fetch: 60
>         max_medias: 10
>     recommendation:
>       show:
>         max_medias: 10
>       movie:
>         max_medias: 10
>     anticipated:
>       anticipated_movie:
>         max_medias: 10
>       anticipated_show:
>         max_medias: 10
> ```
> 
> Can you please tell me what I need to change or add to the existing source files so the integration will actually create the 2 new sensors: `anticipated_movies` and `anticipated_shows`? Thank you for all your help and patience!

Thank YOU, the effort is appreciated a lot, I will look at it tomorrow for sure ;).

downey-lv commented on custom_components/trakt_tv/configuration.py:
```diff
@@ -12,6 +12,13 @@
+            return self.conf["ignore_collected"]
+        except KeyError:
+            return False
+
```
- Move together with the rest of anticipated functions
- `exclude` instead of `ignore` better matches other sources
- Rename to reflect that it's for anticipated only. Maybe `anticipated_exclude_collected`
- Pass identifier to the function to allow separate config for each
- Move the config under anticipated, e.g. `self.conf["anticipated"][identifier]["exclude_collected"]`

downey-lv commented on custom_components/trakt_tv/schema.py:
```diff
@@ -76,4 +80,16 @@ def recommendation_schema() -> Dict[str, Any]:
+    for trakt_kind in ANTICIPATED_KINDS:
+        subschemas[trakt_kind.value.identifier] = {
+            Required("max_medias", default=3): cv.positive_int,
+        }
```
Add exclude_collected to the schema here

downey-lv commented on custom_components/trakt_tv/apis/trakt.py:
```diff
@@ -390,6 +390,61 @@ async def fetch_recommendations(self, configured_kinds: list[TraktKind]):
+                self.fetch_anticipated(
+                    kind.value.path,
+                    configuration.get_anticipated_max_medias(kind.value.identifier),
+                    configuration.ignore_collected
```
Use `configuration.anticipated_exclude_collected(kind.value.identifier)` instead, see my comment on configuration.py

downey-lv commented:
Good job on the improvements! Added some more comments to help you out

> The bad news is that the new sensors are still not being created on my Home Assistant instance after adding this modified trakt integration via config_flow. Only the existing sensors that were there before are being created.

The missing definition of ANTICIPATED_KINDS would definitely break the code, could be the reason why you're not seeing any sensors. 
Once you fix that you can go into http://localhost:8123/config/logs?filter=trakt_tv (change base URL to your instance) and see if anything else is throwing you any errors

mkanet commented:
@downey-lv thank you so much for being so patient with me.  I really appreciate that you guys are so friendly.  I tried my best to make the changes you suggested:  https://github.com/dylandoamaral/trakt-integration/pull/110/commits/c95d4743e635413dd5fb549790328b7ccaf2225d

I also verified that there are no errors:
![image](https://github.com/dylandoamaral/trakt-integration/assets/811008/9c0ef8c5-b2c2-4f69-b89a-121cf24c4c3e)

Unfortunately, the new sensors are still not being created after adding this modified integration via config_flow with no errors at all in home assistant log.  There is something fundamental that I am missing.  Can you please tell me what I should try next?

downey-lv commented on custom_components/trakt_tv/apis/trakt.py:
```diff
@@ -390,6 +390,55 @@ async def fetch_recommendations(self, configured_kinds: list[TraktKind]):
+            "get", f"{path}?limit={limit}&ignore_collected={ignore_collected}"
+        )
+
+    async def fetch_anticipated_medias(self, configured_kinds: "list[TraktKind]"):
```
Remove quotes for configured_kinds

downey-lv commented on custom_components/trakt_tv/sensor.py:
```diff
@@ -51,6 +51,21 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
             )
             sensors.append(sensor)
 
+        # Ensure anticipated sensors are added
```
You've nested it under the first loop which stops if not a BASIC KIND (see line 40), which is why no sensors are created.
It should be separated out at the end of the function
```

    for trakt_kind in TraktKind:
        if trakt_kind not in ANTICIPATED_KINDS:
            continue

        identifier = trakt_kind.value.identifier

        if configuration.anticipated_identifier_exists(identifier):
            sensor = TraktSensor(
                hass=hass,
                config_entry=config_entry,
                coordinator=coordinator,
                trakt_kind=trakt_kind,
                source="anticipated",
                prefix="Trakt Anticipated",
                mdi_icon="mdi:movie" if trakt_kind == TraktKind.ANTICIPATED_MOVIE else "mdi:television",
            )
            sensors.append(sensor)
```

downey-lv commented on custom_components/trakt_tv/models/kind.py:
```diff
@@ -23,6 +23,8 @@ class TraktKind(Enum):
         "only_upcoming", "Only Upcoming", "shows", Show
     )
+    ANTICIPATED_MOVIE = CalendarInformation("anticipated_movie", "Anticipated Movies", "movies/anticipated", Movie)
+    ANTICIPATED_SHOW = CalendarInformation("anticipated_show", "Anticipated Shows", "shows/anticipated", Show)
```
For cleaner code change identifiers from anticipated_XXX to just movie and show, e.g. 
```ANTICIPATED_MOVIE = CalendarInformation("movie", "Anticipated Movies", "movies/anticipated", Movie) ```

Don't forget to update your configuration.yaml as well

downey-lv commented:
I pulled the commit into my development instance, made the changes and the sensors work properly

![image](https://github.com/dylandoamaral/trakt-integration/assets/70920705/a72eb94d-15c5-4600-b2f3-b052ecc54974)
![image](https://github.com/dylandoamaral/trakt-integration/assets/70920705/17663c41-49fc-4fbf-96ad-7e3faaf8bb0c)

Just FYI, I did not test `exclude_collected`

downey-lv commented on custom_components/trakt_tv/models/kind.py:
```diff
@@ -23,6 +23,8 @@ class TraktKind(Enum):
     )
+    ANTICIPATED_MOVIE = CalendarInformation("anticipated_movie", "Anticipated Movies", "movies/anticipated", Movie)
+    ANTICIPATED_SHOW = CalendarInformation("anticipated_show", "Anticipated Shows", "shows/anticipated", Show)
 
```
Remove Anticipated from names, e.g. "Anticipated Movies" to "Movies"
Current set-up will lead to sensors being called "Trakt Anticipated Anticipated Movies"

downey-lv commented:
I pulled the commit, made the changes and the sensors work properly

Just FYI, I did not test `exclude_collected` 

![image](https://github.com/dylandoamaral/trakt-integration/assets/70920705/74cfd113-5cc0-4003-8095-58962cd66a13)

mkanet commented:
thank you.  Yes, it worked for me too.  thank you very much!  Honestly, Ive never used the `exclude_collected` feature before with this integration.  Ill see what kind of effect it will have on the sensor data with it enabled.

mkanet commented:
@downey-lv can you please tell me the minimum code i need to change so that the configuration options are consistent with other configuration settings?  I just realized that `    anticipated:
      anticipated_movie:
        max_medias: 10
      anticipated_show:
        max_medias: 10` is not consistent with other options.  it should be `    anticipated:
      movie:
        max_medias: 10
      show:
        max_medias: 10`  could you please tell me the minimum code changes to achieve that?

downey-lv commented on custom_components/trakt_tv/sensor.py:
```diff
@@ -51,8 +51,10 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
             sensors.append(sensor)
 
-        # Ensure anticipated sensors are added
+    # Added this code
```
Remove your comments

downey-lv commented:
> i need to change so that the configuration options are consistent with other configuration settings?

You can't just have movie and show there, because that would lead to duplicate identifiers. 
You could change the config schema to have movie/show there and then remap them to anticipated_x  in sensor.py instead of having anticipated_x in config. @dylandoamaral what do you prefer?

mkanet commented:
@downey-lv Thank you so much for all your help. As soon as I have some free time I will make the changes and push another commit.  I can't wait.  I really like these new sensors!

mkanet commented:
Hi @downey-lv I tried to do this on my own; but, I'm having a lot of trouble to modify the latest commit I provided in order to support a configuration such as:

```
trakt_tv:
  language: en
  timezone: America/Los_Angeles
  sensors:
    anticipated:
      movie:
        max_medias: 10
      show:
        max_medias: 10
```

I think it would be better than keeping it like:

```
trakt_tv:
  language: en
  timezone: America/Los_Angeles
  sensors:
    anticipated:
      anticipated_movie:
        max_medias: 10
      anticipated_show:
        max_medias: 10
```
I could really use your help to achieve this.

mkanet commented:
Hi @downey-lv  and @dylandoamaral I have made all the changes on my own: 8c6d769e2e64ef085cc06cf9afc77b096c85afd0.  They work correctly as expected for me when using:

```
trakt_tv:
  language: en
  timezone: America/Los_Angeles
  sensors:
    anticipated:
      movie:
        max_medias: 10
      show:
        max_medias: 10
```

It creates the sensors:  `sensor.trakt_anticipated_shows` and `sensor.trakt_anticipated_movies`.  I think I also removed all the temporary comments I had added.

dylandoamaral commented:
> Hi @downey-lv and @dylandoamaral I have made all the changes on my own: [8c6d769](https://github.com/dylandoamaral/trakt-integration/commit/8c6d769e2e64ef085cc06cf9afc77b096c85afd0). They work correctly as expected for me when using:
> 
> ```
> trakt_tv:
>   language: en
>   timezone: America/Los_Angeles
>   sensors:
>     anticipated:
>       movie:
>         max_medias: 10
>       show:
>         max_medias: 10
> ```
> 
> It creates the sensors: `sensor.trakt_anticipated_shows` and `sensor.trakt_anticipated_movies`. I think I also removed all the temporary comments I had added.

Hello @mkanet you also have to merge the `main` branch into your branch and resolve conflicts. I will be in holliday next week but then I will be able to review it !

mkanet commented:
I will submit a new PR for this and close this one.
